File: Winter2022/lab-solutions/java/LabExam/LabExamSkeleton/app/src/main/java/mobiledev/unb/ca/labexam/util/LoadDataTask.java
package mobiledev.unb.ca.labexam.util;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

import mobiledev.unb.ca.labexam.MyAdapter;
import mobiledev.unb.ca.labexam.helper.SharedPreferencesHelper;
import mobiledev.unb.ca.labexam.model.GamesInfo;

public class LoadDataTask {
    private final AppCompatActivity activity;
    private final Context appContext;

    private RecyclerView recyclerView;
    private SharedPreferencesHelper prefsHelper;

    public LoadDataTask(AppCompatActivity activity) {
        this.activity = activity;
        appContext = activity.getApplicationContext();
    }

    public LoadDataTask setRecyclerView(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
        return this;
    }

    public LoadDataTask setSharedPreferencesHelper(SharedPreferencesHelper prefsHelper) {
        this.prefsHelper = prefsHelper;
        return this;
    }

    public void execute() {
        Executors.newSingleThreadExecutor().execute(() -> {
            Handler mainHandler = new Handler(Looper.getMainLooper());

            // TODO
            //  Load the data from the JSON assets file and return the list of cities
            JsonUtils jsonUtils = new JsonUtils(appContext);
            ArrayList<GamesInfo> gamesInfoList = jsonUtils.getHostCities();

            // TODO
            //  Use result to set the adapter for the RecyclerView
            mainHandler.post(() -> setupRecyclerView(gamesInfoList));
        });
    }

    private void setupRecyclerView(List<GamesInfo> gamesInfoList) {
        recyclerView.setAdapter(new MyAdapter(activity, gamesInfoList, prefsHelper));
    }
}

File: Winter2022/lab-solutions/kotlin/LabExam/LabExamSkeleton/app/src/main/java/mobiledev/unb/ca/labexam/util/LoadDataTask.kt
package mobiledev.unb.ca.labexam.util

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import mobiledev.unb.ca.labexam.MyAdapter
import mobiledev.unb.ca.labexam.helper.SharedPreferencesHelper
import mobiledev.unb.ca.labexam.model.GamesInfo
import java.util.concurrent.Executors


class LoadDataTask(private val activity: AppCompatActivity) {
    private val appContext: Context = activity.applicationContext
    private var recyclerView: RecyclerView? = null
    private var sharedPreferencesHelper: SharedPreferencesHelper? = null

    fun setRecyclerView(recyclerView: RecyclerView?): LoadDataTask {
        this.recyclerView = recyclerView
        return this
    }

    fun setSharedPreferencesHelper(sharedPreferencesHelper: SharedPreferencesHelper?): LoadDataTask {
        this.sharedPreferencesHelper = sharedPreferencesHelper
        return this
    }

    fun execute() {
        Executors.newSingleThreadExecutor()
            .execute { val mainHandler = Handler(Looper.getMainLooper())
                // TODO
                //  Load the data from the JSON assets file and return the list of cities
                val jsonUtils = JsonUtils(appContext)
                val gamesInfoList = jsonUtils.hostCities

                // TODO
                //  Use result to set the adapter for the RecyclerView
                mainHandler.post { setupRecyclerView(gamesInfoList!!) }
            }
    }

    private fun setupRecyclerView(gamesInfoList: List<GamesInfo>) {
        recyclerView!!.adapter =
            sharedPreferencesHelper?.let { MyAdapter(activity, gamesInfoList, it) }
    }
}